src/pepr/operator/controllers/keycloak/clients/client-credentials.ts

```diff
@@ -0,0 +1,143 @@
+  cachedToken = null;
+}
+
+export async function credentialsGetAccessToken() {
+  if (cachedToken) {
+    try {
+      const jwt = parseKeycloakToken(cachedToken);
+      if (jwt.exp && jwt.exp > Math.floor(Date.now() / 1000) + 5) return cachedToken;
+    } catch (e) {
+      log.error(e, "Failed to parse cached token");
+      cachedToken = null;
+    }
+  }
+
+  const secret = await K8s(kind.Secret).InNamespace(SECRET_NAMESPACE).Get(SECRET_NAME);
+  if (!secret) throw new Error("Missing secret");
+  const encodedSecret = secret.data?.[UDS_OPERATOR_CLIENT_ID];
+  if (!encodedSecret) throw new Error("Missing client secret");
+
+  const clientSecret = Buffer.from(encodedSecret, "base64").toString("utf-8");
+  const params = new URLSearchParams();
+  params.append("grant_type", "client_credentials");
+  params.append("client_id", UDS_OPERATOR_CLIENT_ID);
+  params.append("client_secret", clientSecret);
+
+  const response = await fetch<KeycloakAccessTokenResponse>(clientCredentialsUrl, {
+    method: "POST",
+    headers: { "Content-Type": "application/x-www-form-urlencoded" },
+    body: params.toString(),
+  });
+  await throwErrorIfNeeded(response);
+  cachedToken = response.data.access_token;
+  return cachedToken;
+}
+
+export async function credentialsCreateOrUpdate(client: Partial<Client>) {
+  log.info(`credentialsCreateOrUpdate: creating/updating client ${JSON.stringify(client)}`);
+  const existingClient = await credentialsGet(client);
+  if (existingClient) {
+    return credentialsUpdate(client);
+  } else {
+    return credentialsCreate(client);
+  }
+}
+
+export async function credentialsCreate(client: Partial<Client>) {
+  log.info(`credentialsCreate: creating client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const response = await fetch(clientsAdminUrl, {
+    method: "POST",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+    body: JSON.stringify(client),
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return credentialsGet(client);
+}
+
+export async function credentialsGet(client: Partial<Client>) {
+  log.info(`credentialsGet: retrieving client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const url = `${clientsAdminUrl}?clientId=${encodeURIComponent(client.clientId!)}`;
+  const response = await fetch<ClientWithId[]>(url, {
+    method: "GET",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return response.data[0];
+}
+
+export async function credentialsUpdate(client: Partial<Client>) {
+  log.info(`credentialsUpdate: updating client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const existing = await credentialsGet(client);
+  const url = `${clientsAdminUrl}/${encodeURIComponent(existing.id!)}`;
```

Comment:
Do we need validation that `existing.id` is non-null? The `!` operator makes me worried this could hide errors if the `existing` response is not what we expect? (same comment for deletion)

Reply:
I don't think so. From the Keycloak side, this needs to be filled in as this is the database id (the primary key, which cannot be null in the database). If we ever see a null there, this needs to be a broader system failure (like Node.js misbehaving or some Linux sockets errors). 

I think we should be safe just to leave it as is.

Reply:
My initial thought was just safe guarding against a bad API response, but might be unnecessary since we're catching !ok responses in the other functions. Okay to leave this as-is for now if you deem it safe - I just tend to prefer the code guaranteeing it rather than relying on an API call to return what we expect. I'm not sure what the result would be here if we called just `${clientsAdminUrl}/` without the ID - if that would return an error and not have unexpected changes then should be fine to allow other error handling to handle this edge case.

My thought was a very lightweight check that guarantees in the typescript code that id is always populated:
```suggestion
  if (!existing || !existing.id) {
    throw new Error(`Failed to retrieve existing client, ${client.clientId}`);
  }
  const url = `${clientsAdminUrl}/${encodeURIComponent(existing.id)}`;
```